protocols/identify/src/identify.rs

```diff
@@ -304,8 +305,14 @@ impl NetworkBehaviour for Identify {
         match event {
             IdentifyHandlerEvent::Identified(info) => {
                 // Replace existing addresses to prevent other peer from filling up our memory.
+                let valid_listen_addrs = info
+                    .listen_addrs
+                    .clone()
```

Comment:
I don't see a reason why a `Multiaddr` that does not fulfill `multiaddr_matches_peer_id` would ever be valid in this context.

With that in mind, how about removing the invalid `Multiaddr`s from `info` directly, thus both (a) only putting valid addresses into `self.discovered_peers` and (b) only emitting valid addresses via `IdentifyEvent::Received`?